### DISCUSSION TITLE
3D?

### DISCUSSION THREAD

**Discussion by NHodgesVFX** [vleue/polyanya#17] [Q&A]
Would it be possible for Vertex and path to take or return either a vec2 or a vec3 so you can navigate in a 3d world, like walking up stairs to a 2nd level?

**Comment by mockersf** [comment]
It depends what you expect in a 3d world:
- navigating as in 2d and just adding a height: already possible in user code
- navigating multiple non overlapping levels: already possible in user code
- navigating multiple overlapping levels: possible but messy in user code, you would have to have "transitions" polygons that wouldn't be overlapped in other levels, and keep track of the current level. This could be added in Polyanya at a small cost
- navigating in a 3d world that still has a plane as a base, and taking height difference in account when calculating cost/shortest path: not possible currently, but I would like to investigate
- navigating in a fully 3d world like a ball or a planet: not possible currently, but I would like to investigate

**Reply by NHodgesVFX** [reply]
![Image of a 3D nav mesh going up some stairs and down, with overlapping areas](https://miro.medium.com/max/563/1*mjFDw7hrGGY6jVqPlZm9vg.jpeg)

Interactive demo: https://godotengine.github.io/godot-demo-projects/3d/navmesh/

The above are examples of what I would like to do which is a 2d navmesh but with height added and overlapping areas although its still just one mesh.

**Reply by mockersf** [reply]
Thanks for the pointer!

This seems to be the third case I mentioned: overlapping levels with non overlapping polygons for transitions.

Polyanya can support this as is, but it requires that you as the caller remember the current level the agent is on, and to transition it yourself when going through one of the transition polygon.

I was mostly missing a mesh to try to do an example of that, I'll try to see if I can reuse the one from the Godot example. No promise on delays though, but that is something I want to make easier to have 👍

**Reply by 0x0539** [reply]
> navigating as in 2d and just adding a height: already possible in user code

I'm trying to figure out how to obtain a 2d mesh from a 3d scene for navigation purposes. My idea is to define a Collider in the shape of a plane (I'm using bevy_xpbd_3d), have it collide with the other Colliders in the scene, extract the ContactManifolds from the Collision events, and use the manifolds to instantiate the 2d mesh somehow. I wonder if anyone has a better idea.
